File: ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/NotaryLookup.java
package net.corda.v5.ledger.common;

import net.corda.v5.base.annotations.DoNotImplement;
import net.corda.v5.base.annotations.Suspendable;
import net.corda.v5.base.types.MemberX500Name;
import net.corda.v5.membership.NotaryInfo;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;

/**
 * Defines a mechanism to allow flows to retrieve the {@link NotaryInfo} in the network.
 * The platform will provide an instance of {@link NotaryLookup} to flows via property injection.
 */
@DoNotImplement
public interface NotaryLookup {

    /**
     * Gets a {@link Collection} of {@link NotaryInfo} services available on the network.
     *
     * @return Returns a {@link Collection} of {@link NotaryInfo} services available on the network.
     */
    @NotNull
    @Suspendable
    Collection<NotaryInfo> getNotaryServices();

    /**
     * Looks up the notary information of a notary by legal name.
     *
     * @param notaryServiceName The {@link MemberX500Name} of the notary service to retrieve.
     * @return Returns the {@link NotaryInfo} that matches the specified notary service name, or null if no such notary exists.
     */
    @Nullable
    @Suspendable
    NotaryInfo lookup(@NotNull MemberX500Name notaryServiceName);

    /**
     * Determines whether the specified virtual node name is a notary, which is defined by the network parameters.
     *
     * @param virtualNodeName The {@link MemberX500Name} to check.
     * @return Returns true if the specified virtual node name is a notary; otherwise, false.
     */
    @Suspendable
    boolean isNotaryVirtualNode(@NotNull MemberX500Name virtualNodeName);
}

File: ledger/ledger-common/src/main/kotlin/net/corda/v5/ledger/common/NotaryLookup.kt
package net.corda.v5.ledger.common

import net.corda.v5.base.annotations.DoNotImplement
import net.corda.v5.base.annotations.Suspendable
import net.corda.v5.base.types.MemberX500Name
import net.corda.v5.membership.NotaryInfo

/**
 * [NotaryLookup] allows flows to retrieve the [NotaryInfo] in the network.
 *
 * The platform will provide an instance of [NotaryLookup] to flows via property injection.
 */
@DoNotImplement
interface NotaryLookup {

    /**
     * Returns a collection of notary services available on the network.
     *
     * @return A collection of the network's Notary services.
     */
    @get:Suspendable
    val notaryServices: Collection<NotaryInfo>

    /**
     * Looks up the notary information of a notary by legal name.
     *
     * @param notaryServiceName The [MemberX500Name] of the notary service
     * 		to retrieve.
     *
     * @return The [NotaryInfo] that matches the input [notaryServiceName], or null if
     * 		no such notary exists.
     */
    @Suspendable
    fun lookup(notaryServiceName: MemberX500Name): NotaryInfo?

    /**
     * Returns true if and only if the given [virtualNodeName] is a notary, which is
     * defined by the network parameters.
     *
     * @param virtualNodeName The [MemberX500Name] to check.
     *
     * @return true if the input [virtualNodeName] is a notary, or false if it is not
     * 		a notary or does not exist.
     */
    @Suspendable
    fun isNotaryVirtualNode(virtualNodeName: MemberX500Name): Boolean
}
